cpp/src/strings/combine.cu

```diff
@@ -0,0 +1,289 @@
+    auto strings_count = first_column->size();
+    if( !std::all_of(strings_columns.begin(),strings_columns.end(),
+        [strings_count] (strings_column_view view) { return strings_count==view.size(); }) )
+    {
+        CUDF_FAIL( "concatenate requires all columns have an equal number of rows");
+    }
+    if( strings_count == 0 )
+        return detail::make_empty_strings_column(mr,stream);
+
+    auto execpol = rmm::exec_policy(stream);
+    if( !separator )
+        separator = "";
+    auto separator_ptr = detail::string_from_host(separator, stream);
+    auto d_separator = *separator_ptr;
+    auto narep_ptr = detail::string_from_host(narep, stream);
+    string_view d_narep(nullptr,0);
+    if( narep_ptr )
+        d_narep = *narep_ptr;
+
+    // Create device views from the strings columns.
+    //
+    // First calculate the size of memory needed to hold the
+    // column_device_views. This is done by calling extent()
+    // for each of the column_views of the strings_columns.
+    size_type views_size_bytes =
+        std::accumulate(strings_columns.begin(), strings_columns.end(), 0,
+            [](size_type init, strings_column_view col) {
+                return init + column_device_view::extent(col.parent());
+            });
+    // Allocate the device memory to be used in the device methods.
+    // We need to pass this down when creating the column_device_views
+    // so they can be resolved to point to any child objects.
+    column_device_view* d_columns;
+    RMM_TRY(RMM_ALLOC(&d_columns, views_size_bytes, stream));
+    column_device_view* d_column = d_columns; // point to the first one
+    // A buffer of CPU memory is created to hold the column_device_view
+    // objects and then copied to device memory at the d_columns pointer.
+    // But each column_device_view instance may have child objects which
+    // require setting an internal device pointer before being copied from
+    // CPU to device.
+    {
+        std::vector<int8_t> h_buffer(views_size_bytes);
+        column_device_view* h_column = reinterpret_cast<column_device_view*>(h_buffer.data());
+        // The beginning of the memory must be the fixed-sized column_device_view
+        // objects in order for d_columns to be used as array. Therefore, any
+        // child data is assigned to the end of this array.
+        int8_t* h_end = (int8_t*)(h_column + num_columns);
+        int8_t* d_end = (int8_t*)(d_column + num_columns);
+        // Create the column_device_view from each column within the CPU memory
+        // array. Any column child data should be copied into h_end and any
+        // internal pointers should be set using d_end.
+        for( auto itr=strings_columns.begin(); itr!=strings_columns.end(); ++itr )
+        {
+            auto col = itr->parent();
+            // convert the column_view into column_device_view
+            new(h_column) column_device_view(col,(ptrdiff_t)h_end,(ptrdiff_t)d_end);
```

Comment:
I'm not seeing how to do this because of the different types involved. The `h_buffer` is an `int8` type and so cannot accept a `column_device_view` directly. The `h_column` is a reinterpret of the beginning of the `h_buffer` as a `column_device_view` only so we can increment through the flat memory with a simple `++` operator.